pkg/ruler/rules/objectclient/rule_store.go

```diff
@@ -103,18 +92,20 @@ func (o *RuleStore) ListRuleGroups(ctx context.Context, userID, namespace string
 		return nil, err
 	}
 
-	groups := []*rules.RuleGroupDesc{}
-	for _, obj := range ruleGroupObjects {
-		level.Debug(util.Logger).Log("msg", "listing rule group", "key", obj.Key)
+	if len(ruleGroupObjects) == 0 {
+		return rules.RuleGroupList{}, nil
+	}
 
-		rg, err := o.getRuleGroup(ctx, obj.Key)
-		if err != nil {
-			level.Error(util.Logger).Log("msg", "unable to retrieve rule group", "err", err, "key", obj.Key)
-			return nil, err
-		}
-		groups = append(groups, rg)
+	groups, err := o.loadRuleGroupsConcurrently(ctx, ruleGroupObjects)
+	if err != nil {
+		return nil, err
+	}
+
+	if _, exists := groups[userID]; !exists {
```

Comment:
No need for extra `if` here... `nil` is a valid empty slice to return. Using just

```
    return groups[userID], nil`
```

or even
```
    return groups[userID], err
```

and removing previous `if err != nil` will work just fine.

Reply:
I tested that in a playground with unsuccessful results https://play.golang.org/p/3-iF20WnP2I

Could be it be that Go's type system coerces nil to the empty return type so that `len` works?

Reply:
Try this instead: https://play.golang.org/p/BGOaFKkQkzR

`nil` interface{} is different than `nil` slice or `nil` map.

Reply:
Thanks! I have addressed the changes.